### PR TITLE
Add maxout op to tf.contrib.layers

### PR DESCRIPTION
**Pull request by gokceneraslan**

### CODE REVIEW

**Comment by gokceneraslan** [comment]
Small demonstration: 

image

**Comment by Mistobaan** [comment]
add a reference to the paper? https://arxiv.org/abs/1302.4389

**Comment by martinwicke** [comment]
@ilblackdragon, can you review?

File: tensorflow/contrib/layers/python/layers/layers.py [hunk 1]
```diff
@@ -1481,6 +1483,57 @@ def max_pool2d(inputs,
...
+      last dimension.
+    outputs_collections: The collections to which the outputs are added.
+    scope: Optional scope for name_scope.
+
+  Returns:
+    A `Tensor` representing the results of the pooling operation.
+
+  Raises:
+    ValueError: if num_units is not multiple of number of features.
+    """
```
Comment (ilblackdragon) [inline tensorflow/contrib/layers/python/layers/layers.py]: the `"""` should be 2 spaces left.

File: tensorflow/contrib/layers/python/layers/layers.py [hunk 2]
```diff
@@ -1481,6 +1483,57 @@ def max_pool2d(inputs,
...
+  with ops.name_scope(scope, 'MaxOut', [inputs]) as sc:
+    inputs = ops.convert_to_tensor(inputs)
+    shape = inputs.get_shape().as_list()
+    if axis is None:
+      # Assume that channel is the last dimension
+      axis = -1
+    num_channels = shape[axis]
```
Comment (ilblackdragon) [inline tensorflow/contrib/layers/python/layers/layers.py]: `num_channels` here can be `None`, e.g. undefined in shape inference. if it's None, you can still implement this via `tf.shape()` and `tf.pack` later.

File: tensorflow/contrib/layers/python/layers/layers.py [hunk 3]
```diff
@@ -1481,6 +1483,57 @@ def max_pool2d(inputs,
...
+    inputs: A Tensor on which maxout will be performed
+    num_units: Specifies how many features will remain after max pooling at the
+      channel dimension. This must be multiple of number of channels.
+    axis: The dimension where max pooling will be performed. Default is the
+      last dimension.
+    outputs_collections: The collections to which the outputs are added.
+    scope: Optional scope for name_scope.
+
+  Returns:
+    A `Tensor` representing the results of the pooling operation.
```
Comment (ilblackdragon) [inline tensorflow/contrib/layers/python/layers/layers.py]: please describe shape of output tensor.

also may be in description add a formula e.g. out[..., i, ..., j] = max(in[...]) (similar how https://www.tensorflow.org/versions/r0.12/api_docs/python/nn.html#conv2d)

File: tensorflow/contrib/layers/python/layers/layers.py [hunk 4]
```diff
@@ -1481,6 +1483,57 @@ def max_pool2d(inputs,
...
+    shape = inputs.get_shape().as_list()
+    if axis is None:
+      # Assume that channel is the last dimension
+      axis = -1
+    num_channels = shape[axis]
+    if num_channels % num_units:
+      raise ValueError('number of features({}) is not '
+                       'a multiple of num_units({})'
+              .format(num_channels, num_units))
+    shape[axis] = -1
```
Comment (ilblackdragon) [inline tensorflow/contrib/layers/python/layers/layers.py]: should it be `= num_units`?
[it's better not to have -1 in reshape, as it's easy to wrongly reshape and not know about it]

**Review by ilblackdragon — CHANGES_REQUESTED** [review]
Thanks for the contribution!

Can you also please add tests. Specifically to check that reshape works correctly for 4-5d tensors with axis in the middle (as it's easy to mix up axis when reshaping).

**Comment by drpngx** [comment]
@gokceneraslan any luck with this?

**Comment by drpngx** [comment]
Closing due to inactivity. Please re-open when updates are available.

**Comment by StefanoD** [comment]
@gokceneraslan Would be interested in this feature! Would be cool, if you would finish this PR! :)

**Comment by tiagofrepereira2012** [comment]
Hi,

I'm interested in this feature.

Is it just a matter of implement some test units and improve the documentation?

**Comment by martinwicke** [comment]
Yes, basically the work to do is to address @ilblackdragon's comments. 

I saw you made a PR, but it was closed, not sure what happened.

**Comment by tiagofrepereira2012** [comment]
I don't know what happened to; I'll reopen it

**Pull request closed without merging**